zerver/lib/outgoing_webhook.py

```diff
@@ -0,0 +1,72 @@
+from __future__ import absolute_import
+from typing import Any, Iterable, Dict, Optional, Set, Tuple
+from six import text_type
+
+import requests
+import sys
+import inspect
+from six.moves import urllib
+
+from zerver.models import get_outgoing_webhook_bot_profile, Realm
+
+from straight.plugin import load
+
+class OutgoingWebhookBotInterface(object):
+    email = None # type: text_type
+    full_name = None # type: text_type
+
+    def __init__(self):
+        self.post_url = None # type: text_type
+        self.service_api_key = None # type: text_type
+
+    def process_command(self, command):
+        raise NotImplementedError()
+
+    def process_response(self, response_json):
+        raise NotImplementedError()
+
+    def handle_failure(self, status_code, json_response):
+        raise NotImplementedError()
+
+class OutgoingWebhookBot():
+
+    def __init__(self, interface):
+        self.post_url = interface.post_url
+        self.service_api_key = interface.service_api_key
+        self.process_command = interface.process_command
+        self.process_response = interface.process_response
+        self.handle_failure = interface.handle_failure
+
+    def trigger(self, command, timeout=None):
+        # type: (text_type, float) -> text_type
+        http_operation, relative_url_path, options = self.process_command(command)
+        final_url = urllib.parse.urljoin(self.post_url, relative_url_path)
+
+        options['timeout'] = timeout
+        response = requests.request(http_operation, final_url, **options)
+        if response.status_code == requests.codes.ok:
+            message = self.process_response(response.json())
+            return message
+        else:
+            return self.handle_failure(response.status_code, response.json())
+
+def load_available_bots():
+    # type: () -> Dict[text_type, Any]
+    modules = load("zerver.outgoing_webhooks")
+    class_members = [inspect.getmembers(module, inspect.isclass) for module in modules] # type: List[List[Tuple[str, Any]]]
+    if len(class_members) > 0:
+        class_member_list = reduce(list.__add__, class_members) # type: List[Tuple[str, Any]]
+    else:
+        class_member_list = []
+    bot_classes = {cls.email:cls for cls_name,cls in class_member_list if cls.email is not None}
+    return bot_classes
+
+def get_outgoing_webhook_bot_handler(bot_email, realm):
+    # type: (text_type, Realm) -> OutgoingWebhookBot
+    bot_profile = get_outgoing_webhook_bot_profile(bot_email, realm)
+    bot_interface = available_outhook_bots[bot_email](post_url=bot_profile.post_url,
+                                                      service_api_key=bot_profile.service_api_key)
+    bot_handler = OutgoingWebhookBot(bot_interface)
+    return bot_handler
```

Comment:
I'm not sure I understand how this file is intended to interact with the classes in zerver/models.py.